src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
```

Comment:
I'm a bit confused here.
We can step here only if there is no `retryTemplate`.
But at the same time `recoveryCallback` is used only if `retryTemplate` is performed.
So, this attributes populatation looks slightly as a dead code.

On the other hand, when we have `retryTemplate`, we populate context via `open(RetryContext context, RetryCallback<T, E> callback)` and we seems for me we don't set these attributes in that case.

Since it is so messed (at least for me), I would like to know the algorithm how it is going to work now.

Thanks.